Poker game step3

1Consumption opened the pull request:
* 고민한 점
  * Card type, CardDeck type의 인스턴스가 같은지 아닌지를 어떻게 비교할 것인가?
    * shuffle() 후 카드가 섞였는지를 확인하기 위해 CardDeck을 비교해야할 필요가 있었습니다. 이를 위해 CardDeck 배열 내의 Card를 하나하나 대조해보려 했으나, CardDeck 내의 Card는 어떻게 비교할지에 대해 고민에 빠지게 되었습니다. 이를 위해 Equatable 프로토콜을 채택해주었고, Card의 경우 인스턴스를 문자열로 반환한 값이 같으면 같다고 판단하였습니다. CardDeck의 경우는 먼저 가지고 있는 카드의 수가 같은지를 비교하였고, 같다면 배열 내의 값의 순서를 비교하였습니다.
  * struct의 property를 변경하려면 어떻게 해야하는가?
    * mutating을 사용하였습니다. 구조체는 선언 이후에 값을 변경하지 못하는데, 값을 변경해줘야 할 경우에는 변경된 값을 바탕으로 새로운 구조체를 만들어 변수에 할당해주는 방식으로 동작한다고 알고있습니다. 하지만 이 부분의 경우 명확히 알지 못하기에 조금 더 공부를 하겠습니다.

1Consumption committed:
CardDeck 구조체 파일 생성

카드 객체 인스턴스를 포함하는 CardDeck 구조체

1Consumption committed:
CardDeck의 크기 반환하는count 변수 생성

Read only computed properties로 생성하여 외부에서 값 변경 못하도록 설정

1Consumption committed:
cardSet 변수 private 선언, count 변수를 함수로 수정

cardSet변수를 외부에서 접근하지 못하게 private으로 선언
cardSet의 값을 구조체 내부에서 지정해주기 위해 mutating 키워드를 사용하여 cardSet의 값을 초기화하는 setUpCardSet 메소드 생성

1Consumption committed:
shuffle() 메소드 및 테스트 코드 생성

기본으로 제공하는 shuffle() 메소드를 이용해서 카드 셔플, 카드가 두장 이상일 때만 셔플하게 함

CardGameAppTests 클래스에 shuffle 여부를 검사하는 testShuffle 메소드 생성
class를 비교하기 위해 Card와 CardDeck 클래스에 Equatable 프로토콜 채용해 줌

1Consumption committed:
removeOne() 메소드 및 테스트 메소드 추가

카드가 없을 때 removeOne()을 호출할 경우를 고려하여 반환형을 Card? 로두었음. guard문을 활용하여 카드가 없는 경우를 체크하고, 없으면 nil 반환

testRemoveOne() 메소드에서 카드가 없는경우 nil을 반환하는지를 체크하고 카드가 있다면, Card 타입을 반환하는지, 반환하고 난 후 카드의 수가 하나 줄어들었는지 체크

1Consumption committed:
reset() 메소드 및 테스트 함수 추가

카드 덱을 생성하는 setUpCardDeck()함수를 통해 새로운 cardDeck 생성

1Consumption committed:
readme 수정

추가한 기능 정리

1Consumption committed:
readme 수정

구조체와 클래스의 차이
fisher yates 알고리즘 작동  방식 정리

godrm commented on CardGameApp/CardGameApp/CardDeck.swift:
```diff
@@ -0,0 +1,63 @@
+            return false
+        }
+    }
+}
```
원래 선언하는 struct CardDeck 을 먼저 쓰고, extension CardDeck을 아래 쓰는 게 위에서 아래로 읽는 흐름상 좋습니다.

1Consumption replied:
넵 extension을 어디에 써야하는지 확인했습니다.

godrm commented on CardGameApp/CardGameApp/CardDeck.swift:
```diff
@@ -0,0 +1,63 @@
+extension CardDeck:Equatable{
+    static func == (lhs: CardDeck, rhs: CardDeck) -> Bool {
+        if lhs.count() == rhs.count(){
+            for index in 0..<lhs.cardSet.count{
```
extension 경우 file private처럼 동작하기 때문에 내부 속성에 접근 가능합니다.
배열 같은 걸 탐색할 때 index 기반보다는 Sequence 형태로 접근하는 게 가능하도록 구현하는 게 좋습니다.
lhs 와 rhs 탐색할 때 `for item in lhs` 처럼 탐색하도록 개선해보세요. 
아예 탐색하지 않고 전체를 비교하는 방법도 가능합니다.

1Consumption replied:
내부 속성에 접근하여 cardset 전체를 한번에 비교하도록 수정하였습니다.

godrm commented on CardGameApp/CardGameApp/card.swift:
```diff
@@ -14,6 +14,15 @@ extension Card:CustomStringConvertible{
 
+extension Card:Equatable{
+    static func == (lhs: Card, rhs: Card) -> Bool {
+        if lhs.description == rhs.description{
```
description을 비교하기 보다는 내부 속성을 그대로 비교하세요.
그리고 이런 경우는 return lhs.suit == rhs.suit 이런 식으로 return 뒤에 조건을 써도 됩니다.

1Consumption replied:
넵 내부속성 비교하였습니다.

godrm reviewed (commented):
다음 피드백을 확인해보세요.

1Consumption committed:
enum rawValue type 수정, 클래스 비교 시 내부 프로퍼티로 비교

코드내에서 extension의 위치를 struct 아래로 수정
Rank, Suit enum의 rawValue를 String 으로 수정하여 불필요한 변환 과정 줄임
Card 인스턴스를 비교할 때 기존 description을 기준으로 비교하였지만 내부 프로퍼티에 접근하여 rank와 suit 값을 각각 비교 후 결과 리턴
CardDeck 인스턴스를 비교할 때 기존 인덱스 기준으로 하나하나 비교하였지만 내부 프로퍼티에 접근하여 cardSet 시퀀스 전체를 비교하도록 수정 후 결과 리턴

1Consumption commented:
피드백 반영하여 수정하였습니다. 확인 부탁드립니다!

godrm commented:
수정 사항도 커밋을 작게 나누는 습관을 가지세요

godrm commented on CardGameApp/CardGameApp/CardDeck.swift:
```diff
@@ -61,3 +46,12 @@ struct CardDeck{
+        guard lhs.cardSet == rhs.cardSet else {
+            return false
+        }
+        return true
```
guard else 구문이 필요없이 return 해보세요

1Consumption replied:
수정하였습니다. 불필요한 guard 문을 제거할 수 있었습니다.

1Consumption committed:
Card 비교 부분 수정

불필요한 guard문 제거
